sdk/proto/provider.proto

```diff
@@ -40,6 +40,9 @@ service ResourceProvider {
     // of responses.
     rpc StreamInvoke(InvokeRequest) returns (stream InvokeResponse) {}
 
+    // Call dynamically executes a method in the provider associated with a component resource.
+    rpc Call(CallRequest) returns (CallResponse) {}
```

Comment:
@pgavlin, I'm curious to hear your thoughts on `Call`. The original design was for methods to be implemented with `Invoke` (and we'd augment `Invoke` with the additional information needed to be able to create resources). However, @lukehoban and I were wondering if it'd be better to have a separate RPC, since we'd be adapting the raw RPC values into the Pulumi SDK model in each language, just like `Construct`. Which then leaves `Invoke` for leaf functions that just provide raw compute / data sources.

I've added `Call` in this PR, and think it's probably best to have it, but part of me still feels a little unsure. On the one hand, I do think it simplifies the implementation in the providers: it's more clear when implementing `Call` that the context and values of the call are that of the Pulumi SDK model. On the other hand, we're now bifurcating functions listed in the schema where some are implemented with `Invoke` (leaf functions that _don't_ accept inputs and don't create resources, i.e. raw compute / data sources) and others are implemented with `Call` (methods & leaf functions that do accept inputs and create resources).

If we're going to keep `Call` one question I have is what do we do about mocking. Our mocking APIs for `Invoke` are actually named `Call` (e.g. [Node.js](https://github.com/pulumi/pulumi/blob/d525a8b3318bdc2f9292bc06c106ec8ce6ee0b92/sdk/nodejs/runtime/mocks.ts#L88), [Python](https://github.com/pulumi/pulumi/blob/d525a8b3318bdc2f9292bc06c106ec8ce6ee0b92/sdk/python/lib/pulumi/runtime/mocks.py#L100), [Go](https://github.com/pulumi/pulumi/blob/d525a8b3318bdc2f9292bc06c106ec8ce6ee0b92/sdk/go/pulumi/mocks.go#L19), [.NET](https://github.com/pulumi/pulumi/blob/d525a8b3318bdc2f9292bc06c106ec8ce6ee0b92/sdk/dotnet/Pulumi/Testing/IMocks.cs#L27)), so I don't think we'd be able to distinguish these unless we came up with a different name for `Call` (or used a different name for the mock API). I was thinking we'd just have both `Invoke` and `Call` pass through to the mock's `Call` (similar to how both `registerResource` and `readReasource` are passed through to the mock [`newResource`](https://github.com/pulumi/pulumi/blob/d525a8b3318bdc2f9292bc06c106ec8ce6ee0b92/sdk/nodejs/runtime/mocks.ts#L96)).

Reply:
Briefly answered earlier.

I do think that we should keep `Call`: there is additional info that we need in this method that we really don't need for PoD functions (e.g. input dependencies, monitor info), and I agree that this simplifies the implementation.

> I was thinking we may be able to get by having both Invoke and Call pass through to the mock's Call (similar to how both registerResource and readReasource are passed through to the mock newResource).

This would be my preference for now. It might be the case that in the future we need to separate remote component resource registration and `Call` out into their own mocks, as the shapes of the inputs are a bit different (i.e. the user's mocks might actually want to see `Input`s rather than a bunch of data + dependency info).